internal/image/newest_build_selector.go

```diff
@@ -49,6 +52,118 @@ func (n *newestBuildSelector) Select(ctx context.Context) (*Image, error) {
 
 	ctx = logging.ContextWithLogger(ctx, logger)
 
+	images, err := n.selectImages(ctx)
+	if err != nil || len(images) == 0 {
+		return nil, err
+	}
+
+	if n.platform == nil {
+		image := images[0]
+		logger.WithFields(log.Fields{
+			"tag":    image.Tag,
+			"digest": image.Digest.String(),
+		}).Trace("found image")
+		return &image, nil
+	}
+
+	tag := images[0].Tag
+	digest := images[0].Digest
+	image, err := n.repoClient.getImageByDigest(ctx, digest, n.platform)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving image with digest %q: %w", digest.String(), err)
+	}
+	if image == nil {
+		logger.Tracef(
+			"image with digest %q was found, but did not match platform constraint",
+			digest.String(),
+		)
+		return nil, nil
+	}
+	image.Tag = tag
+
+	logger.WithFields(log.Fields{
+		"tag":    image.Tag,
+		"digest": image.Digest.String(),
+	}).Trace("found image")
+	return image, nil
+}
+
+func (n *newestBuildSelector) Discover(ctx context.Context) ([]Image, error) {
+	logger := logging.LoggerFromContext(ctx).WithFields(log.Fields{
+		"registry":            n.repoClient.registry.name,
+		"image":               n.repoClient.image,
+		"selectionStrategy":   SelectionStrategyNewestBuild,
+		"platformConstrained": n.platform != nil,
+		"discoveryLimit":      n.discoveryLimit,
+	})
+	logger.Trace("discovering images")
+
+	ctx = logging.ContextWithLogger(ctx, logger)
+
+	images, err := n.selectImages(ctx)
+	if err != nil || len(images) == 0 {
+		return nil, err
+	}
+
+	limit := n.discoveryLimit
+	if limit == 0 || limit > len(images) {
+		limit = len(images)
+	}
+
+	if n.platform == nil {
+		for _, image := range images[:limit] {
+			logger.WithFields(log.Fields{
+				"tag":    image.Tag,
+				"digest": image.Digest.String(),
+			}).Trace("discovered image")
+		}
+		logger.Tracef("discovered %d images", limit)
+		return images[:limit], nil
+	}
+
+	// TODO(hidde): this is unnecessarily inefficient, as we're fetching the
+	// image _again_ to check if it matches the platform constraint. We should
+	// refactor this to avoid fetching the image twice.
```

Comment:
Agree we could do better here, but fwiw, the second retrieval of each is by digest and will always be a cache hit.

Reply:
Rephrased the comment a bit to highlight this.